File: app/src/main/java/com/bwei/p2p/WelcomeActivity.java
package com.bwei.p2p;

import android.annotation.SuppressLint;
import android.app.AlertDialog;
import android.app.ProgressDialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.os.Environment;
import android.os.Handler;
import android.os.Message;
import android.util.Log;
import android.view.Window;
import android.view.WindowManager;
import android.view.animation.AccelerateInterpolator;
import android.view.animation.AlphaAnimation;
import android.widget.RelativeLayout;
import android.widget.TextView;
import android.widget.Toast;

import com.bwei.base.ActivityInstanceManager;
import com.bwei.base.BaseActivity;
import com.bwei.base.bean.Index;
import com.bwei.base.bean.UpdateInfo;

import java.io.File;

public class WelcomeActivity extends BaseActivity {

    private static final int TO_MAIN = 1;
    private RelativeLayout rlWelcome;
    private TextView tvWelcomeVersion;
    @SuppressLint("HandlerLeak")
    private Handler handler = new Handler() {
        @Override
        public void handleMessage(Message msg) {
            switch (msg.what) {
                case TO_MAIN:
                    finish();
                    startActivity(new Intent(WelcomeActivity.this, MainActivity.class));
                    break;
            }

        }
    };


    protected void initView() {
        rlWelcome = (RelativeLayout) findViewById(R.id.rl_welcome);
        tvWelcomeVersion = (TextView) findViewById(R.id.tv_welcome_version);
    }

    @Override
    protected void initDate() {
        //将当前的activity添加到ActivityManager中
        ActivityInstanceManager.addActivity(this);
        //提供启动动画
        setAnimation();
        boolean connected = isConnected();
        if (connected){
            //有网络
            CacheManager.getInstance().registerGetDateListener(new CacheManager.onGetDateListener() {
                @Override
                public void onGetData(Index index) {
                }

                @Override
                public void onGetUpDate(UpdateInfo upDate) {
                    //更新信息
                    String version = getVersion();
                    tvWelcomeVersion.setText(version);
                    //比较服务器获取的最新的版本跟本应用的版本是否一致
                    if(version.equals(upDate.version)){
                        Toast.makeText(WelcomeActivity.this, "当前应用已经是最新版本", Toast.LENGTH_SHORT).show();
                        toMain();
                    }else{
                        new AlertDialog.Builder(WelcomeActivity.this)
                                .setTitle("下载最新版本")
                                .setMessage(upDate.desc)
                                .setPositiveButton("确定", new DialogInterface.OnClickListener() {
                                    @Override
                                    public void onClick(DialogInterface dialog, int which) {
                                        //下载服务器保存的应用数据
                                        downloadApk();
                                    }
                                })
                                .setNegativeButton("取消", new DialogInterface.OnClickListener() {
                                    @Override
                                    public void onClick(DialogInterface dialog, int which) {
                                        toMain();
                                    }
                                })
                                .show();
                    }
                }
            });
        }else {
            Toast.makeText(this,"当前没有网络连接",Toast.LENGTH_SHORT).show();
        }
        toMain();


    }
    private ProgressDialog dialog;
    private File apkFile;
    private void downloadApk() {
        //初始化水平进度条的dialog
        dialog = new ProgressDialog(this);
        dialog.setProgressStyle(ProgressDialog.STYLE_HORIZONTAL);
        dialog.setCancelable(false);
        dialog.show();
        //初始化数据要保持的位置
        File filesDir;
        if(Environment.getExternalStorageState().equals(Environment.MEDIA_MOUNTED)){
            filesDir = this.getExternalFilesDir("");
        }else{
            filesDir = this.getFilesDir();
        }
        apkFile = new File(filesDir,"update.apk");
    }

    private String getVersion() {
        String version="未知版本";
        PackageManager packageName = getPackageManager();
        try {
            PackageInfo packageInfo = packageName.getPackageInfo(getPackageName(), 0);
            version = packageInfo.versionName;
            Log.i("SSSSS", "getVersion: "+version);

        } catch (PackageManager.NameNotFoundException e) {
            throw new RuntimeException(e.getMessage());
        }
        return version;

    }

    @Override
    protected int getLayoutId() {
        // 去掉窗口标题
        requestWindowFeature(Window.FEATURE_NO_TITLE);
        // 隐藏顶部的状态栏
        getWindow().addFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN);
        return R.layout.activity_welcome;
    }
    private void toMain() {
        long currentTime = System.currentTimeMillis();
        handler.sendEmptyMessageDelayed(TO_MAIN, 3000);
    }
    private void setAnimation() {
        AlphaAnimation alphaAnimation = new AlphaAnimation(0, 1);//0：完全透明  1：完全不透明
        alphaAnimation.setDuration(3000);
        alphaAnimation.setInterpolator(new AccelerateInterpolator());//设置动画的变化率

        //启动动画
        rlWelcome.startAnimation(alphaAnimation);

    }


    @Override
    public void onConnected() {
        if (!isConnected()) {
            Toast.makeText(this, "当前网络没有连接", Toast.LENGTH_SHORT).show();
            return;
        }
        Toast.makeText(this, "当前网络连接正常，获取数据", Toast.LENGTH_SHORT).show();
    }

    @Override
    public void onDisConnected() {
        Toast.makeText(this,"当前网络没有连接",Toast.LENGTH_SHORT).show();

    }
}

File: app/src/main/java/com/example/p2pdemo/activity/WelcomeActivity.kt
package com.example.p2pdemo.activity

import android.content.Intent
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.view.WindowManager
import android.view.animation.AlphaAnimation
import com.example.modulebase.BaseActivity
import com.example.p2pdemo.R
import kotlinx.android.synthetic.main.activity_welcome.*
import java.util.*

class WelcomeActivity : BaseActivity() {
    override fun getLayoutId(): Int {
        return R.layout.activity_welcome
    }



    override fun initData() {
        setAnimator()
    }
    override fun flagFullScreen() {
        //activity全屏显示 隐藏顶部状态栏
        window.addFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN)
    }
    //设置入场动画 并动画结束后跳转
    private fun setAnimator(){
        //0完全透明到1完全不透明
        var alphaAnimation = AlphaAnimation(0.2f,1f)
        //动画时间
        alphaAnimation.duration = 3000

        //动画监听
//        alphaAnimation.setAnimationListener(object : Animation.AnimationListener{
//            override fun onAnimationRepeat(animation: Animation?) {
//                //在动画结束时跳转到主界面并销毁当前页面
//                startActivity(Intent(this@WelcomeActivity,MainActivity::class.java))
//                finish()
//            }
//
//            override fun onAnimationStart(animation: Animation?) {
//            }
//
//            override fun onAnimationEnd(animation: Animation?) {
//            }
//
//        })

        //使用时间戳控制跳转主页面并销毁当前页面
        var count = 0
        var timer = Timer()
        timer.schedule(object : TimerTask() {
            override fun run() {
             count++
                if (count == 4){
                    startActivity(Intent(this@WelcomeActivity,
                        MainActivity::class.java))
                    finish()
                }
            }
        },0,1000)


        //给需要的View启动动画
    rl_welcome.startAnimation(alphaAnimation)

    }
}
